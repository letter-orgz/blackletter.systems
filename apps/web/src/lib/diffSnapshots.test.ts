import { describe, it, expect } from 'vitest';

function simpleDiff(oldStr: string, newStr: string, options: { ignoreWhitespace?: boolean } = {}) {
  const oldLines = oldStr.split('\n');
  const newLines = newStr.split('\n');
  const maxLen = Math.max(oldLines.length, newLines.length);
  const result: string[] = [];
  for (let i = 0; i < maxLen; i++) {
    const a = oldLines[i] ?? '';
    const b = newLines[i] ?? '';
    const aComp = options.ignoreWhitespace ? a.trim() : a;
    const bComp = options.ignoreWhitespace ? b.trim() : b;
    if (aComp === bComp) {
      result.push(`  ${b}`);
    } else {
      if (a) result.push(`- ${a}`);
      if (b) result.push(`+ ${b}`);
    }
  }
  return result.join('\n');
}

describe('diff snapshots', () => {
  it('captures small diff', () => {
    const before = 'alpha\nbeta\ngamma';
    const after = 'alpha\nbeta changed\ngamma';
    const diff = simpleDiff(before, after);
    expect(diff).toMatchSnapshot();
  });

  it('captures large diff', () => {
    const before = Array.from({ length: 20 }, (_, i) => `line ${i}`).join('\n');
    const after = Array.from({ length: 20 }, (_, i) =>
      i === 10 ? `line ${i} modified` : `line ${i}`
    ).join('\n');
    const diff = simpleDiff(before, after);
    expect(diff).toMatchSnapshot();
  });
});

describe('comment thread snapshots', () => {
  it('captures unresolved and resolved threads', () => {
    const unresolved = {
      id: 1,
      resolved: false,
      comments: ['Needs changes', 'Please update'],
    };
    const resolved = {
      ...unresolved,
      resolved: true,
      resolution: 'Fixed in latest commit',
    };
    expect({ unresolved, resolved }).toMatchSnapshot();
  });
});

describe('whitespace options', () => {
  it('captures diffs with and without whitespace ignored', () => {
    const before = 'a\nb\nc';
    const after = 'a\nb  \n c';
    const diffDefault = simpleDiff(before, after);
    const diffIgnored = simpleDiff(before, after, { ignoreWhitespace: true });
    expect({ diffDefault, diffIgnored }).toMatchSnapshot();
  });
});
