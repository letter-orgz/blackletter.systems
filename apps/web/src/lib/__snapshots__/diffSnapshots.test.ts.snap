// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`comment thread snapshots > captures unresolved and resolved threads 1`] = `
{
  "resolved": {
    "comments": [
      "Needs changes",
      "Please update",
    ],
    "id": 1,
    "resolution": "Fixed in latest commit",
    "resolved": true,
  },
  "unresolved": {
    "comments": [
      "Needs changes",
      "Please update",
    ],
    "id": 1,
    "resolved": false,
  },
}
`;

exports[`diff snapshots > captures large diff 1`] = `
"  line 0
  line 1
  line 2
  line 3
  line 4
  line 5
  line 6
  line 7
  line 8
  line 9
- line 10
+ line 10 modified
  line 11
  line 12
  line 13
  line 14
  line 15
  line 16
  line 17
  line 18
  line 19"
`;

exports[`diff snapshots > captures small diff 1`] = `
"  alpha
- beta
+ beta changed
  gamma"
`;

exports[`whitespace options > captures diffs with and without whitespace ignored 1`] = `
{
  "diffDefault": "  a
- b
+ b  
- c
+  c",
  "diffIgnored": "  a
  b  
   c",
}
`;
